### PR TITLE
Switch to dto-decorators for implementation.

### PR DESCRIPTION
**Pull request by jessemyers**
The underlying implementation of `@hippo-oss/nest-dto` has become too complex and too difficult
to extend. We've now re-designed the implementation so that:

 - The basic decorator types and composition operations live in `@hippo-oss/dto-decorators`.
 - The `class-transformer` and `class-validator` flavors live in `@hippo-oss/class-decorators.`
 - The `@nestjs/swagger` implementation lives in `@hippo-oss/openapi-decorators`.

This rewrite makes this library mostly a shim around the other libraries, plus some utilities
for modifying (pick/omit) decorated types and for testing OpenAPI schemas.

This change also deprecated the `basic` flavor and the `@IsArray` decorator.

In exchange, we get a more maintainable code base, the ability to run a subset of the decorators
nicely (e.g. excluding `@nestjs/swagger` in React apps), and a much simpler process for defining
new decorator implementations.

### CODE REVIEW

File: src/flavors/__tests__/__snapshots__/openapi.test.ts.snap [hunk 1]
```diff
@@ -209,9 +210,6 @@ Object {
             "type": "string",
           },
         },
-        "required": Array [
```
Comment (jessemyers) [inline src/flavors/__tests__/__snapshots__/openapi.test.ts.snap:212]: This change is due to a bug in the previous declaration. A "nullable" field was not declared as such.

File: src/flavors/__tests__/__snapshots__/openapi.test.ts.snap [hunk 2]
```diff
@@ -356,6 +354,7 @@ Array [
     "constraints": Object {
       "isArray": "requiredPrimitiveArrayValue must be an array",
       "isDefined": "requiredPrimitiveArrayValue should not be null or undefined",
+      "isNumber": "each value in requiredPrimitiveArrayValue must be a number conforming to the specified constraints",
```
Comment (jessemyers) [inline src/flavors/__tests__/__snapshots__/openapi.test.ts.snap:357]: This change is due to the previous implementation using `@IsArray` and not giving `class-validator` enough info.

File: src/flavors/__tests__/fixtures.ts [hunk 3]
```diff
@@ -23,46 +22,57 @@ export function createFixtures(flavor: Flavor): Fixture {
...
+        @IsString({
+            optional: true,
+        })
         optionalStringValue?: string;
     }
 
     class NestedNullableExample {
-        @IsString()
+        @IsString({
+            nullable: true,
```
Comment (jessemyers) [inline src/flavors/__tests__/fixtures.ts:39]: This was missing `nullable`

File: src/flavors/__tests__/fixtures.ts [hunk 4]
```diff
@@ -23,46 +22,57 @@ export function createFixtures(flavor: Flavor): Fixture {
...
-        @IsString()
+        @IsString({
+        })
         requiredStringValue!: string;
     }
 
     class NestedOptionalExample {
-        @IsString()
+        @IsString({
+            optional: true,
```
Comment (jessemyers) [inline src/flavors/__tests__/fixtures.ts:32]: This was missing `optional`

File: src/flavors/openapi.ts [hunk 5]
```diff
@@ -1,41 +1,20 @@
...
+import { composeDecoratorFactories } from '@hippo-oss/dto-decorators';
+import { CLASS_DECORATORS } from '@hippo-oss/class-decorators';
+import { OPENAPI_DECORATORS } from '@hippo-oss/openapi-decorators';
 
-const initializers = [
-    initializeSwagger,
-    initializeStrictTransformer,
-    initializeValidator,
-];
+export const DECORATORS = composeDecoratorFactories([
```
Comment (jessemyers) [inline src/flavors/openapi.ts:5]: Compose the `openapi` and `class-*` decorators.

File: src/flavors/strict.ts [hunk 6]
```diff
@@ -1,38 +1,18 @@
...
-    IsUUIDRecipe,
-} from '../recipes';
+import { composeDecoratorFactories } from '@hippo-oss/dto-decorators';
+import { CLASS_DECORATORS } from '@hippo-oss/class-decorators';
 
-const initializers = [
-    initializeStrictTransformer,
-    initializeValidator,
-];
+export const DECORATORS = composeDecoratorFactories([
```
Comment (jessemyers) [inline src/flavors/strict.ts:4]: I don't actually have to compose here, but I like symmetry.

File: src/operators/index.ts [hunk 7]
```diff
@@ -1,4 +1,3 @@
 export * from './derive.operator';
 export * from './omit.operator';
-export * from './null-to-undefined.operator';
```
Comment (jessemyers) [inline src/operators/index.ts:3]: This was only used by a `@Transform` call that is now in `class-decorators`. The other operators are all used for shared types magic to pick/omit decorated fields.

**Review by ranhalprin — APPROVED** [review]

File: src/flavors/__tests__/fixtures.ts [hunk 8]
```diff
@@ -23,46 +22,57 @@ export function createFixtures(flavor: Flavor): Fixture {
     }
 
     class NestedRequiredExample {
-        @IsString()
+        @IsString({
+        })
```
Comment (IodizedGabe) [inline src/flavors/__tests__/fixtures.ts:26]: Hrm, it would be a bit nicer to not have to pass in empty args to here.

**Review by IodizedGabe — APPROVED** [review]

File: src/flavors/__tests__/fixtures.ts (hunk 8 above)
Comment (jessemyers) [reply src/flavors/__tests__/fixtures.ts:26]: It is difficult to define such things generically due to some decorators having mandatory options.

**Pull request merged**